pkg/util/admission/granter.go

```diff
@@ -418,157 +446,128 @@ func (tg *tokenGranter) grantKind() grantKind {
-	// matter that kvGranter internally uses both slots and tokens.
-	return slot
+func (sg *kvStoreTokenGranter) grantKind() grantKind {
+	return token
 }
 
-func (sg *kvGranter) tryGet() bool {
-	return sg.coord.tryGet(KVWork)
+func (sg *kvStoreTokenGranter) tryGet(count int64) bool {
+	return sg.coord.tryGet(KVWork, count)
 }
 
-func (sg *kvGranter) tryGetLocked() grantResult {
-	if sg.usedSlots < sg.totalSlots || sg.skipSlotEnforcement {
-		if !sg.ioTokensEnabled || sg.availableIOTokens > 0 {
-			sg.usedSlots++
-			if sg.usedSlotsMetric != nil {
-				sg.usedSlotsMetric.Update(int64(sg.usedSlots))
-			}
-			if sg.ioTokensEnabled {
-				sg.availableIOTokens--
-				if sg.availableIOTokens == 0 {
-					sg.exhaustedStart = timeutil.Now()
-				}
-			}
-			return grantSuccess
-		}
-		return grantFailLocal
+func (sg *kvStoreTokenGranter) tryGetLocked(count int64) grantResult {
+	if sg.availableIOTokens > 0 {
+		sg.subtractTokens(count, false)
+		return grantSuccess
 	}
-	return grantFailDueToSharedResource
+	return grantFailLocal
 }
 
-func (sg *kvGranter) returnGrant() {
-	sg.coord.returnGrant(KVWork)
+func (sg *kvStoreTokenGranter) returnGrant(count int64) {
+	sg.coord.returnGrant(KVWork, count)
 }
 
-func (sg *kvGranter) returnGrantLocked() {
-	sg.usedSlots--
-	if sg.usedSlots < 0 {
-		panic(errors.AssertionFailedf("used slots is negative %d", sg.usedSlots))
-	}
-	if sg.usedSlotsMetric != nil {
-		sg.usedSlotsMetric.Update(int64(sg.usedSlots))
-	}
+func (sg *kvStoreTokenGranter) returnGrantLocked(count int64) {
+	sg.subtractTokens(-count, false)
 }
 
-func (sg *kvGranter) tookWithoutPermission() {
-	sg.coord.tookWithoutPermission(KVWork)
+func (sg *kvStoreTokenGranter) tookWithoutPermission(count int64) {
+	sg.coord.tookWithoutPermission(KVWork, count)
 }
 
-func (sg *kvGranter) tookWithoutPermissionLocked() {
-	sg.usedSlots++
-	if sg.usedSlotsMetric != nil {
-		sg.usedSlotsMetric.Update(int64(sg.usedSlots))
-	}
-	if sg.ioTokensEnabled {
-		sg.availableIOTokens--
-		if sg.availableIOTokens == 0 {
-			sg.exhaustedStart = timeutil.Now()
+func (sg *kvStoreTokenGranter) tookWithoutPermissionLocked(count int64) {
+	sg.subtractTokens(count, false)
+}
+
+func (sg *kvStoreTokenGranter) subtractTokens(count int64, forceTickMetric bool) {
+	avail := sg.availableIOTokens
+	sg.availableIOTokens -= count
+	if count > 0 && avail > 0 && sg.availableIOTokens <= 0 {
+		// Transition from > 0 to <= 0.
+		sg.exhaustedStart = timeutil.Now()
+	} else if count < 0 && avail <= 0 && (sg.availableIOTokens > 0 || forceTickMetric) {
+		// Transition from <= 0 to > 0, or forced to tick the metric. The latter
+		// ensures that if the available tokens stay <= 0, we don't show a sudden
+		// change in the metric after minutes of exhaustion (we had observed such
+		// behavior prior to this change).
+		now := timeutil.Now()
+		exhaustedMicros := now.Sub(sg.exhaustedStart).Microseconds()
+		sg.ioTokensExhaustedDurationMetric.Inc(exhaustedMicros)
+		if sg.availableIOTokens <= 0 {
```

Comment:
@sumeerbhola would there be any interest in backporting just this change from `== 0` to `<= 0` to v21.2?

The `ioTokensExhaustedDurationMetric` appears somewhat busted on v21.2 because every time the function containing this code (`setAvailableIOTokensLocked` in v21.2) is called when `sg.availableIOTokens` is less than 0, it re-adds `now() - sg.exhaustedStart` to the metric without resetting the value of `sg.exhaustedStart`.

On a node that's constantly overloaded, it ends up with a `ioTokensExhaustedDurationMetric` that increases roughly quadratically, such that taking its rate gives a linearly increasing graph over time, which doesn't seem to be the metric's intent? (especially given that this commit appears to have fixed it)

Reply:
Sorry, I mean v22.1, not v21.2

Reply:
yo @a-robinson, nice see you around! Few of us are out on PTO and others are swarming to shove stuff in for the impending release cut. Filed https://github.com/cockroachdb/cockroach/issues/86963 to make sure we fix it in 22.1, what you're saying looks right to me.

Reply:
Thanks @irfansharif ! Clearly no rush since it's just a metric, just thought I'd mention it.